### PR TITLE
Add WaitGroup to sendCommand

### LINKED ISSUE #37: Update sendCommand to use WaitGroup
If `instanceIDs` is large, there's a reasonable chance that some command executions will complete sooner than others, and it's _possible_ that some might take so much time that the service-side record of the command might be garbage-collected before we retrieve the execution status.

```suggestion
	// Wait for the sent commands to complete
	wg := sync.WaitGroup{}
	for _, v := range instanceIDs {
		wg.Add(1)
		go func(instanceID string) {
			  ssmClient.WaitUntilCommandExecuted(&ssm.GetCommandInvocationInput{
				  CommandId:  aws.String(commandID),
				  InstanceId: aws.String(instanceID),
			  })
			  wg.Done()
		}(aws.StringValue(v))
	}
	wg.Wait()
```

_Originally posted by @samuelkarp in https://github.com/bottlerocket-os/bottlerocket-ecs-updater/pull/34#discussion_r610932709_

### PR DESCRIPTION
**Pull request by WilboMo**
**Issue number:**
Closes #37 

**Description of changes:**
This change adds WaitGroup to `sendCommand` to prevent edge cases where long running commands
get garbage collected before execution status is retrieved.

**Testing done:**
ran `make test` : 

```
PASS
ok      github.com/bottlerocket-os/bottlerocket-ecs-updater     49.596s
```

Executed changes against an ECS test cluster and verified successful update actions on test cluster: 

```
2021/07/13 15:27:54 Container instance "arn:aws:ecs:us-west-2:518928554640:container-instance/updater-test-cluster/43dd5fb5949d477c943770caf794990d" updated to version "1.1.2"
2021/07/13 15:27:54 Instance {`i-071b02327101e91d2` `arn:aws:ecs:us-west-2:518928554640:container-instance/updater-test-cluster/43dd5fb5949d477c943770caf794990d` `1.0.5`} updated successfully!

2021/07/13 15:29:16 Container instance "arn:aws:ecs:us-west-2:518928554640:container-instance/updater-test-cluster/f973cfed0d8f42c1b61b66cb8e03cf84" updated to version "1.1.3"
2021/07/13 15:29:16 Instance {`i-065a986fb41aaee3b` `arn:aws:ecs:us-west-2:518928554640:container-instance/updater-test-cluster/f973cfed0d8f42c1b61b66cb8e03cf84` `1.0.5`} updated successfully!
```
Above instances were destroyed after testing. 

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: updater/aws.go [hunk 1]
```diff
@@ -432,24 +433,35 @@ func (u *updater) sendCommand(instanceIDs []string, ssmDocument string) (string,
...
+			err = u.ssm.WaitUntilCommandExecutedWithContext(aws.BackgroundContext(), &ssm.GetCommandInvocationInput{
+				CommandId:  aws.String(commandID),
+				InstanceId: aws.String(instanceID),
+			},
+				request.WithWaiterMaxAttempts(waiterMaxAttempts),
+				request.WithWaiterDelay(request.ConstantWaiterDelay(waiterDelay)))
+			if err != nil {
+				errChan <- fmt.Errorf("error encountered while awaiting document %q execution for instance: %q: %s", ssmDocument, instanceID, err)
+				u.logCommmandOutput(commandID, instanceID)
+			}
```
Comment (srgothi92) [inline updater/aws.go]: nit: I think it's better to print the error message as and when it occurs and collect just the error in channel to decide how many errors occurred and print the last error message.
```suggestion
			if err != nil {
			        errChan <- err
			        log.Printf("Error encountered while awaiting document %q execution for instance: %q: %s", ssmDocument, v, err)
				u.logCommmandOutput(commandID, instanceID)
			}
```

File: updater/aws_test.go [hunk 2]
```diff
@@ -90,7 +90,8 @@ func TestSendCommandSuccess(t *testing.T) {
 	commandID, err := u.sendCommand(instances, "test-doc")
 	require.NoError(t, err)
 	assert.EqualValues(t, "command-id", commandID)
-	assert.Equal(t, instances, waitInstanceIDs)
+	assert.Contains(t, waitInstanceIDs, "inst-id-2")
+	assert.Contains(t, waitInstanceIDs, "inst-id-1")
```
Comment (srgothi92) [inline updater/aws_test.go]: https://pkg.go.dev/github.com/stretchr/testify/assert#ElementsMatch
Same suggestion can be applied to similar assertions in the file .
```suggestion
	assert.ElementsMatch(t, instances, waitInstanceIDs)
```

Comment (WilboMo) [reply updater/aws_test.go]: Oh nice, I somehow didn't come across/notice `ElementsMatch` when looking through the different assertions for something better than `Contains`. Thanks for pointing that out!

File: updater/aws_test.go [hunk 3]
```diff
@@ -189,10 +192,11 @@ func TestSendCommandWaitSuccess(t *testing.T) {
 		commandID, err := u.sendCommand(instances, "test-doc")
 		require.NoError(t, err)
 		assert.Equal(t, "command-id", commandID)
-		assert.Equal(t, expectedFailInstances, failedInstanceIDs, "should match instances for which wait fail")
+		assert.Contains(t, expectedFailInstances, "inst-id-1", "should match instance for which wait fails")
```
Comment (gthao313) [inline updater/aws_test.go]: According to shailesh's suggestion above, if we can use `ElementsMatch` here as well?  :slightly_smiling_face:

File: updater/aws_test.go [hunk 4]
```diff
@@ -1027,7 +1027,7 @@ func TestUpdateInstanceErr(t *testing.T) {
 			containerInstanceID: "cont-inst-id",
 		})
 		require.Error(t, err)
-		assert.ErrorIs(t, err, waitExecErr)
+		assert.Contains(t, err.Error(), "failed to wait ssm execution complete")
```
Comment (srgothi92) [inline updater/aws_test.go]: Does it gives error without this change? I am trying to understand how the code change to affect this assertion.

Comment (WilboMo) [reply updater/aws_test.go]: Before I changed the error handling to pass `err` to the channel I had it wrapping the error which did cause the test to fail, so I changed it to `Contains` like other cases which validate wrapped errors. This escaped me when I made the last revision. I'll switch it back to `ErrorIs` since thats working now.

Comment (srgothi92) [reply updater/aws_test.go]: Similarly, there is one more similar assertion which might wanna fix.

Comment (WilboMo) [reply updater/aws_test.go]: For sure, already planning on changing it too :)

**Review by gthao313 — APPROVED** [review]

**Review by srgothi92 — APPROVED** [review]

File: updater/aws.go [hunk 5]
```diff
@@ -432,24 +433,36 @@ func (u *updater) sendCommand(instanceIDs []string, ssmDocument string) (string,
...
-		}
+		wg.Add(1)
+		go func(instanceID string) {
+			defer wg.Done()
+			err = u.ssm.WaitUntilCommandExecutedWithContext(aws.BackgroundContext(), &ssm.GetCommandInvocationInput{
+				CommandId:  aws.String(commandID),
+				InstanceId: aws.String(instanceID),
+			},
+				request.WithWaiterMaxAttempts(waiterMaxAttempts),
+				request.WithWaiterDelay(request.ConstantWaiterDelay(waiterDelay)))
```
Comment (samuelkarp) [inline updater/aws.go:449]: I'd prefer that we remove the constant delay here.  If there are concurrent calls and the calls end up being throttled, exponential backoff and a bit of jitter are both useful for smoothing the call rate.
```suggestion
```

**Review by samuelkarp — COMMENTED** [review]
The code mostly looks good, with one minor comment about retries.  However, can you verify that the rate of concurrent calls to `GetCommandInvocation` is going to be below any default service limit from SSM?  The code here will generate up to 50 calls (the page size used by the caller of `sendCommand`) at roughly the same time; if that's too many and would trigger throttles we should instead limit the number of concurrent calls we generate.

File: updater/aws.go (hunk 5 above)
Comment (WilboMo) [reply updater/aws.go:449]: I verified that a max concurrency of 50 will not be an issue for the `GetCommandInvocation` API.

Comment (srgothi92) [reply updater/aws.go:449]: If we remove `waiterDelay` overall wait time for SSM command can change because by default the [waiter Delay is 5 seconds](https://github.com/aws/aws-sdk-go/blob/v1.38.20/service/ssm/waiters.go#L32) and our `waiterDelay` is 15 seconds. Would "exponential backoff and a bit of jitter" help keep the same max wait time as before ?

Comment (samuelkarp) [reply updater/aws.go:449]: @srgothi92 Thanks for pointing that out; it looks like the waiter delay is constant by default.  My concern with a constant delay applies to both the default behavior and our 15-second change, so...  @WilboMo Can you put back the line I suggested you take out?  My apologies for the churn here...

(The concern with a constant delay and concurrent calls is that all the calls hit the SSM API in a very short window.  Since @WilboMo confirmed that 50 is not a problem, I'm comfortable having that be the case for now.  Longer term, I'd like to see something like a `JitteredWaiterDelay` to spread out the load a bit more.)

Comment (WilboMo) [reply updater/aws.go:449]: Sure thing!

**Review by samuelkarp — APPROVED** [review]

File: updater/aws.go [hunk 6]
```diff
@@ -432,24 +433,36 @@ func (u *updater) sendCommand(instanceIDs []string, ssmDocument string) (string,
...
+			if err != nil {
+				errChan <- err
+				log.Printf("error encountered while awaiting document %q execution for instance: %q: %s", ssmDocument, instanceID, err)
```
Comment (srgothi92) [inline updater/aws.go]: nit: Sorry, Noticed it just now.
```suggestion
				log.Printf("Error encountered while awaiting document %q execution for instance: %q: %s", ssmDocument, instanceID, err)
```

**Review by srgothi92 — APPROVED** [review]
